src/adapter/src/coord/caught_up.rs

```diff
@@ -385,4 +419,139 @@ impl Coordinator {
             ctx.trigger.fire();
         }
     }
+
+    /// Analyzes replica status history to detect replicas that are
+    /// crash-looping or OOM-looping.
+    ///
+    /// A replica is considered problematic if it has multiple failures in a
+    /// short time window. This includes crashes, OOM kills, or other non-ready
+    /// states that repeat frequently.
+    async fn analyze_replica_looping(&self, now: EpochMillis) -> BTreeSet<ReplicaId> {
+        // Look back 1 day for patterns.
+        let lookback_window: u64 = Duration::from_secs(24 * 60 * 60)
+            .as_millis()
+            .try_into()
+            .expect("fits into u64");
+        let min_timestamp = now.saturating_sub(lookback_window);
+        let min_timestamp_dt = mz_ore::now::to_datetime(min_timestamp);
+
+        // Get the replica status collection GlobalId
+        let replica_status_item_id = self
+            .catalog()
+            .resolve_builtin_storage_collection(&MZ_CLUSTER_REPLICA_STATUS_HISTORY);
+        let replica_status_gid = self
+            .catalog()
+            .get_entry(&replica_status_item_id)
+            .latest_global_id();
+
+        // Acquire a read hold to determine the as_of timestamp for snapshot_and_stream
+        let read_holds = self
+            .controller
+            .storage_collections
+            .acquire_read_holds(vec![replica_status_gid])
+            .expect("can't acquire read hold for mz_cluster_replica_status_history");
+        let read_hold = if let Some(read_hold) = read_holds.into_iter().next() {
+            read_hold
+        } else {
+            // Collection is not readable anymore, but we return an empty set
+            // instead of panicing.
+            return BTreeSet::new();
+        };
+
+        let as_of = read_hold
+            .since()
+            .iter()
+            .next()
+            .cloned()
+            .expect("since should not be empty");
+
+        let mut replica_statuses_stream = self
+            .controller
+            .storage_collections
+            .snapshot_and_stream(replica_status_gid, as_of)
+            .await
+            .expect("can't read mz_cluster_replica_status_history");
+
+        let mut replica_problem_counts: BTreeMap<ReplicaId, u32> = BTreeMap::new();
+
+        while let Some((source_data, _ts, diff)) = replica_statuses_stream.next().await {
+            // Only process inserts (positive diffs)
+            if diff <= 0 {
+                continue;
+            }
```

Comment:
`snapshot_and_stream`'s doc comment says "The output is __not__ consolidated.", so I think we should consolidate it ourselves to not need to handle negative diffs here. (I guess we could even soft_assert that there are no negative diffs after consolidation.)